src/Microsoft.AspNetCore.OData/Extensions/Endpoint/ODataEndpointRouteValueTransformer.cs

```diff
@@ -0,0 +1,201 @@
+                {
+                    // Set all the properties we need for routing, querying, formatting
+                    IODataFeature odataFeature = httpContext.ODataFeature();
+                    odataFeature.Path = path;
+                    odataFeature.RouteName = routeName;
+                    odataFeature.IsEndpointRouting = true; // mark as Endpoint routing
+
+                    // Noted: we inject the ActionSelector and use it to select the best OData action.
+                    // In .NET 5 or later, this maybe change.
+                    RouteContext routeContext = new RouteContext(httpContext);
+                    var condidates = _selector.SelectCandidates(routeContext);
+                    var actionDescriptor = _selector.SelectBestCandidate(routeContext, condidates);
+                    ControllerActionDescriptor controllerActionDescriptor = actionDescriptor as ControllerActionDescriptor;
+                    if (controllerActionDescriptor != null)
+                    {
+                        RouteValueDictionary newValues = new RouteValueDictionary();
+                        foreach (var item in values)
+                        {
+                            newValues.Add(item.Key, item.Value);
+                        }
+
+                        foreach (var item in routeContext.RouteData.Values)
+                        {
+                            newValues[item.Key] = item.Value;
+                        }
+
+                        newValues["controller"] = controllerActionDescriptor.ControllerName;
+                        newValues["action"] = controllerActionDescriptor.ActionName;
+                        newValues["odataPath"] = oDataPathValue;
+
+                        // Noted, here's a working around for mulitiple actions in same controller.
+                        // For example, we have two "Get" methods in same controller, in order to help "EndpointSelector" 
+                        // to select the correct Endpoint, we save the ActionDescriptor value into ODataFeature.
+                        odataFeature.ActionDescriptor = controllerActionDescriptor;
+
+                        return new ValueTask<RouteValueDictionary>(newValues);
+                    }
+                }
+                else
+                {
+                    // The request doesn't match this route so dispose the request container.
+                    request.DeleteRequestContainer(true);
+                }
+            }
+
+            return new ValueTask<RouteValueDictionary>(result: null);
+        }
+    }
+
+    internal class ODataDynamicControllerEndpointMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
+    {
+        private readonly EndpointMetadataComparer _comparer;
+        public ODataDynamicControllerEndpointMatcherPolicy(EndpointMetadataComparer comparer)
+        {
+            if (comparer == null)
+            {
+                throw new ArgumentNullException(nameof(comparer));
```

Comment:
@xuzhg Maybe use `throw Error.ArgumentNull(...)` for consistency

Reply:
This class is removed and replaced using "ODataEndpointSelectorPolicy"

---
In reply to: [384446213](https://github.com/OData/WebApi/pull/2035#discussion_r384446213) [](ancestors = 384446213)